Implement a constructor for nested_tensor that is similar to torch.tensor()

ani300 opened the pull request:
Summary: This diff merges both previous implementations of constructors for nested tensors, the one from lists of tensors and the one with arbitrary python lists, adn implements it in pytorch core so no extensions are needed to construct NT.

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
+                length = np.random.randint(low=1, high=max_seq_len)
+            row = list(np.random.randint(low=0, high=VOCAB_SIZE, size=(length,)))
+            data.append(row)
+            nested_tensor_ref_list.append(torch.Tensor(row))
```
nit: Please us `torch.tensor`

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
+        nested_tensor = torch.nested.nested_tensor(data, dtype=torch.int)
+        nested_tensor_list = nested_tensor.unbind()
+        for id in range(batch_size):
+            assert torch.equal(
```
Please use self.assertEqual

cpuhrsch commented on test/test_nestedtensor.py:
```diff
@@ -93,6 +97,79 @@ def random_nt(device, dtype, num_tensors, max_dims, min_dims=None):
+        max_seq_len=st.integers(min_value=2, max_value=4),
+    )
+    @settings(deadline=None)
+    def test_3d_nested_tensor(self, batch_size: int, max_seq_len: int):
```
Above comments also apply to the tests below

cpuhrsch commented on torch/csrc/api/include/torch/nested.h:
```diff
@@ -1,7 +1,10 @@
+#include <torch/csrc/api/include/torch/detail/TensorDataContainer.h>
+#include <algorithm>
+#include <stdexcept>
+#include "ATen/core/ATen_fwd.h"
```
Please use angle brackets

cpuhrsch commented on torch/csrc/api/include/torch/nested.h:
```diff
@@ -12,22 +15,42 @@ namespace nested {
+    if (nested_tensor_data[0].is_init_list()) {
+      // TODO: Implement conversion from InitList to NT for C++,
+      // link into existing infrastructure at torch/csrc/utils/nested.cpp
+      throw std::runtime_error(
```
Please use `TORCH_CHECK`

cpuhrsch commented on torch/csrc/api/include/torch/nested.h:
```diff
@@ -36,10 +59,10 @@ inline Tensor nested_tensor(
-    TensorList list,
-    c10::optional<ScalarType> dtype = c10::nullopt,
-    c10::optional<Device> device = c10::nullopt) {
+inline at::Tensor as_nested_tensor(
```
Why are these changes needed?

ani300 replied:
The compiler started complaining for some reason.

cpuhrsch commented on torch/csrc/utils/nested.cpp:
```diff
@@ -0,0 +1,222 @@
+namespace torch {
+namespace utils {
+
+long compute_total_size_and_sizes_tensor(
```
Is it possible to rely on `_nested_tensor_from_tensor_list` for all the dtype, device, etc. checks?

ani300 replied:
Looking at the implementation for both torch::tensor() and _nested_tensor_from_tensor_list, we do the same amount of checks

cpuhrsch replied:
Sure, but this could deduplicate code

cpuhrsch commented on torch/csrc/api/include/torch/nested.h:
```diff
@@ -12,34 +15,71 @@ namespace nested {
+inline at::Tensor nested_tensor(
+    at::ArrayRef<detail::TensorDataContainer> nested_tensor_data,
+    const at::TensorOptions& options = {}) {
+  if (nested_tensor_data.size() > 0) {
```
nit: I don't think you should need this check
```
>>> torch.nested.nested_tensor([])
nested_tensor([

])
```

ani300 replied:
this is for the C++ path, given that python now takes python vars it needs different impls (or at least torch::tensor does so)

mikaylagawarecki commented on torch/nested/__init__.py:
```diff
@@ -137,4 +113,34 @@ def as_nested_tensor(tensor_list: List[Tensor], dtype: Optional[DType] = None, d
+
+nested_tensor = _add_docstr(
+    _nested.nested_tensor,
+    r"""
```
nit: when using `_add_docstr`, in order to render the arguments for the function a line containing the signature of the function needs to be added here (see `to_padded_tensor` above for an example)

(the preview for the docs is available under helpful links [in the pytorch-bot comment on this PR](https://github.com/pytorch/pytorch/pull/88213#issuecomment-1298866390) if you would like to look at the rendering!)

Commit:
Implement a constructor for nested_tensor that is similar to torch.tensor() (#88213)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88213

This diff merges both previous implementations of constructors for nested tensors, the one from lists of tensors and the one with arbitrary python lists, adn implements it in pytorch core so no extensions are needed to construct NT.

Test Plan:
All nested tensor tests pass successfully. Specifically:

`buck test mode/dev-nosan //caffe2/test:nested`

Reviewed By: cpuhrsch, mikekgfb

Differential Revision: D40878312

fbshipit-source-id: e9f36fa66287299efc144c45b65a8f6cbe248ca3

ani300 commented:
@pytorchbot merge -g

cpuhrsch commented:
@pytorchbot merge -g
